src/cpu/pred/gshare.cc

```diff
@@ -0,0 +1,180 @@
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * * Author: Karan Pathak
+ */
+
+/* @file
+ * Implementation of a bi-mode branch predictor
+ */
+
+#include "cpu/pred/gshare.hh"
+
+#include "base/bitfield.hh"
+#include "base/intmath.hh"
+
+namespace gem5
+{
+
+namespace branch_prediction
+{
+
+GshareBP::GshareBP(const GshareBPParams &params)
+    : BPredUnit(params),
+      globalHistoryReg(params.numThreads, 0),
+      globalHistoryBits(ceilLog2(params.globalPredictorSize)),
+      globalPredictorSize(params.globalPredictorSize),
+      globalCtrBits(params.globalCtrBits),
+      globalCtrs(globalPredictorSize, SatCounter8(globalCtrBits))
+{
+
+    if (!isPowerOf2(globalPredictorSize))
+        fatal("Invalid global history predictor size.\n");
+
+    historyRegisterMask = mask(globalHistoryBits);
+    globalHistoryMask = globalPredictorSize - 1;
+    takenThreshold = (1ULL << (globalCtrBits - 1)) - 1;
+}
+
+/*
+ * For an unconditional branch we set its history such that
+ * everything is set to taken. I.e., its choice predictor
+ * chooses the taken array and the taken array predicts taken.
+ */
+void
+GshareBP::uncondBranch(ThreadID tid, Addr pc, void * &bp_history)
+{
+    BPHistory *history = new BPHistory;
+    history->globalHistoryReg = globalHistoryReg[tid];
+    history->finalPred = true;
+    bp_history = static_cast<void*>(history);
+}
+
+void
+GshareBP::updateHistories(ThreadID tid, Addr pc, bool uncond,
+                         bool taken, Addr target,
+                         const StaticInstPtr &inst, void * &bp_history)
+{
+    assert(uncond || bp_history);
+    if (uncond) {
+        uncondBranch(tid, pc, bp_history);
+    }
+    updateGlobalHistReg(tid, taken);
+}
+
+
+void
+GshareBP::squash(ThreadID tid, void * &bp_history)
+{
+    BPHistory *history = static_cast<BPHistory*>(bp_history);
+    globalHistoryReg[tid] = history->globalHistoryReg;
+
+    delete history;
+    bp_history = nullptr;
+}
+
+/*
+ * Here we lookup the actual branch prediction. A hash of
+ * the global history register and a branch's PC is used to
+ * index into the counter, which both present a prediction.
+ */
+bool
+GshareBP::lookup(ThreadID tid, Addr branchAddr, void * &bp_history)
+{
+    unsigned globalHistoryIdx = (((branchAddr >> instShiftAmt)
+                                ^ globalHistoryReg[tid])
+                                & globalHistoryMask);
+
+    assert(globalHistoryIdx < globalPredictorSize);
+
+    bool finalPrediction = globalCtrs[globalHistoryIdx] > takenThreshold;
```

Comment:
Local variable should be snake_case
```suggestion
    bool final_prediction = globalCtrs[globalHistoryIdx] > takenThreshold;
```